fdbserver/QuietDatabase.actor.cpp

```diff
@@ -670,6 +671,60 @@ ACTOR Future<Void> reconfigureAfter(Database cx,
 	return Void();
 }
 
+struct QuietDatabaseChecker {
+	double start = now();
+	constexpr static double maxDDRunTime = 1000.0;
+
+	struct Impl {
+		double start;
+		std::string const& phase;
+		std::vector<std::string> failReasons;
+
+		Impl(double start, const std::string& phase) : start(start), phase(phase) {}
+
+		template <class T, class Comparison = std::less_equal<>>
+		Impl& add(BaseTraceEvent& evt,
+		          const char* name,
+		          T value,
+		          T expected,
+		          Comparison const& cmp = std::less_equal<>()) {
+			std::string k = fmt::format("{}Gate", name);
+			evt.detail(name, value).detail(k.c_str(), expected);
+			if (!cmp(value, expected)) {
+				failReasons.push_back(name);
+			}
+			return *this;
+		}
+
+		bool success() {
+			bool timedOut = now() - start > maxDDRunTime;
+			if (!failReasons.empty()) {
+				std::string traceMessage = fmt::format("QuietDatabase{}Fail", phase);
+				std::string reasons = fmt::format("{}", fmt::join(failReasons, ", "));
+				TraceEvent(timedOut ? SevError : SevWarnAlways, traceMessage.c_str())
+				    .detail("Reasons", reasons)
+				    .detail("FailedAfter", now() - start)
+				    .detail("Timeout", maxDDRunTime);
+				if (timedOut) {
+					// this bool is just created to make the assertion more readable
+					bool ddGotStuck = true;
+					// This assertion is here to make the test fail more quickly. If quietDatabase takes this
+					// long without completing, we can assume that the test will eventually time out. However,
+					// time outs are more annoying to debug. This will hopefully be easier to track down.
+					ASSERT(!ddGotStuck);
```

Comment:
Adding this comment again since the previous one is now outdated:

Should this ASSERT be something we only do in simulation? I'm not sure what effect this will have on any non-simulated tests that run consistency check.

Reply:
That's a good point. I believe that if you run `fdbserver -r consistencycheck` it won't actually run `QuietDatabase` as we do this in `tester.actor.cpp`:

```
		options.push_back_deep(options.arena(),
		                       KeyValueRef(LiteralStringRef("performQuiescentChecks"), LiteralStringRef("false")));
```

But I think you're still right because we might want to change this in the future, or we might want to add another role that runs QuietDatabase in the future.